crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -7100,20 +7173,74 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_subscript_expression(&mut self, subscript: &ast::ExprSubscript) -> Type<'db> {
+        let ast::ExprSubscript {
+            value,
+            slice,
+            range: _,
+            ctx,
+        } = subscript;
+
+        match ctx {
+            ExprContext::Load => self.infer_subscript_load(subscript),
+            ExprContext::Store | ExprContext::Del => {
+                let value_ty = self.infer_expression(value);
+                let slice_ty = self.infer_expression(slice);
+                self.infer_subscript_expression_types(value, value_ty, slice_ty);
+                Type::Never
+            }
+            ExprContext::Invalid => {
+                let value_ty = self.infer_expression(value);
+                let slice_ty = self.infer_expression(slice);
+                self.infer_subscript_expression_types(value, value_ty, slice_ty);
+                Type::unknown()
+            }
+        }
+    }
+
+    fn infer_subscript_load(&mut self, subscript: &ast::ExprSubscript) -> Type<'db> {
         let ast::ExprSubscript {
             range: _,
             value,
             slice,
             ctx: _,
         } = subscript;
+        let db = self.db();
+        let value_ty = self.infer_expression(value);
+
+        if let Ok(expr) = PlaceExpr::try_from(subscript) {
+            // Type narrowing based on assignment to a subscript expression is generally unsound,
+            // because `__getitem__`/`__setitem__` of a user-defined class does not guarantee that the passed value is stored and can be retrieved as is.
+            // Currently, we only perform assignment-based narrowing on a few built-in classes and their subclasses
+            // where we are confident that this kind of narrowing can be performed soundly. This is the same approach as pyright.
+            let safe_mutable_classes = [
+                KnownClass::List.to_instance(db),
+                KnownClass::Dict.to_instance(db),
+                KnownClass::Bytearray.to_instance(db),
+                Type::KnownInstance(KnownInstanceType::ChainMap),
+                Type::KnownInstance(KnownInstanceType::Counter),
+                Type::KnownInstance(KnownInstanceType::DefaultDict),
+                Type::KnownInstance(KnownInstanceType::Deque),
+                Type::KnownInstance(KnownInstanceType::OrderedDict),
+                Type::KnownInstance(KnownInstanceType::TypedDict),
+            ];
+            if safe_mutable_classes
+                .iter()
+                .any(|class| value_ty.is_assignable_to(db, *class))
+            {
```

Comment:
See https://github.com/mtshiba/ruff/blob/narrow_by_assignment/crates/ty_python_semantic/resources/mdtest/narrow/assignment.md#subscript for why narrowing should not be performed in these cases.

Perhaps other classes can be considered safe, but we cannot expect that all standard library types will be soundly implemented (e.g. `email.message.EmailMessage`).